File: cpp/dataStructures/advanced/Pathset.cpp
/** @file PathSet.cpp
 *
 *  @author Jon Turner
 *  @date 2011
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */
#include "PathSet.h"

#define left(x) pnode[x].left
#define right(x) pnode[x].right
#define p(x) pnode[x].p
#define dcost(x) pnode[x].dcost
#define dmin(x) pnode[x].dmin

namespace grafalgo {

/** Constructor for PathSet class.
 *  @param size defines the index range for the constructed object.
 *  @param pathVals is a vector of integer values, which is defined for
 *  each path in the set; that is if u is the handle of some path,
 *  then pathVals[u] will be an integer which the application using
 *  the PathSet has defined for the path; since the PathSet object
 *  may change the handle of a path as a side effect of any operation,
 *  the PathSet object updates pathVals whenever the handle for a path
 *  changes
 */
PathSet::PathSet(int size, int *pathVals) : Adt(size), pvals(pathVals) {
	makeSpace(size);
}

/** Destructor for PathSet class. */
PathSet::~PathSet() { freeSpace(); }

/** Allocate and initialize space for PathSet.
 *  @param size is number of index values to provide space for
 */
void PathSet::makeSpace(int size) {
	try {
		pnode = new PathNode[size+1];
	} catch (std::bad_alloc e) {
		string s = "makeSpace:: insufficient space for "
			    + to_string(size) + "index values";
		throw OutOfSpaceException(s);
	}
	nn = size; clear();
}

/** Free dynamic storage used by PathSet. */
void PathSet::freeSpace() {
	delete [] pnode;
}

/** Reinitialize data structure, creating single node trees. */
void PathSet::clear() {
	for (index i = 0; i <= n(); i++) {
		left(i) = right(i) = p(i) = dcost(i) = dmin(i) = 0;
	}
}

/** Resize a PathSet object, discarding old value.
 *  @param size is the size of the resized object.
 */
void PathSet::resize(int size) {
	freeSpace();
	try { makeSpace(size); } catch(OutOfSpaceException e) {
		string s = "PathSet::resize::" + e.toString();
		throw OutOfSpaceException(s);
	}
}

/** Expand the space available for this ojbect.
 *  Rebuilds old value in new space.
 *  @param size is the size of the expanded object.
 */
void PathSet::expand(int size) {
	if (size <= n()) return;
	PathSet old(this->n(),pvals); old.copyFrom(*this);
	resize(size); this->copyFrom(old);
}

/** Copy another object to this one.
 *  @param source is object to be copied to this one
 */
void PathSet::copyFrom(const PathSet& source) {
	if (&source == this) return;
	if (source.n() > n()) resize(source.n());
	else clear();

	for (index x = 1; x <= n(); x++) {
		pnode[x] = source.pnode[x];
		pvals[x] = source.pvals[x];
	}
}

/** Perform a splay operation on the binary search tree representing a path.
 *  @param x is a node in some path; the operation does a splay operation
 *  that moves x to the root of the search tree that represents the path
 *  containing x
 */
index PathSet::splay(index x) {
	while (p(x) != 0) splaystep(x);
	return x;
}

/** Perform a single splay step.
 *  @param x is a node in some path
 */
void PathSet::splaystep(index x) {
        index y = p(x);
        if (y == 0) return;
        index z = p(y);
        if (x == left(left(z)) || x == right(right(z)))
                rotate(y);
        else if (z != 0) // x is "inner grandchild"
                rotate(x);
        rotate(x);
}

/** Perform a rotation in a search tree representing a path.
 *  @param x is a node in some path; the operation performs a rotation
 *  at the parent of x, moving x up into its parent's position.
 */
void PathSet::rotate(index x) {
        index y = p(x); if (y == 0) return;
        index a, b, c;
        if (x == left(y)) { a = left(x);  b = right(x); c = right(y); }
        else              { a = right(x); b = left(x);  c = left(y);  }

	// do the rotation
        p(x) = p(y);
             if (y == left(p(y)))  left(p(x)) = x;
        else if (y == right(p(y))) right(p(x)) = x;
        if (x == left(y)) {
                left(y) = right(x);
                if (left(y) != 0) p(left(y)) = y;
                right(x) = y;
        } else {
                right(y) = left(x);
                if (right(y) != 0) p(right(y)) = y;
                left(x) = y;
        }
        p(y) = x;

	// update dmin, dcost values
        dmin(a) += dmin(x); dmin(b) += dmin(x);

        dcost(x) = dcost(x) + dmin(x);
        cost dmx = dmin(x);
        dmin(x) = dmin(y);

        dmin(y) = dcost(y);
        if (b != 0) dmin(y) = min(dmin(y),dmin(b)+dmx);
        if (c != 0) dmin(y) = min(dmin(y),dmin(c));
        dcost(y) = dcost(y) - dmin(y);

        dmin(b) -= dmin(y); dmin(c) -= dmin(y);

	pvals[x] = pvals[y]; // ensures that root node always has path value
}

/** Return the canonical element of some path.
 *  @param i is a node in some path
 *  @return the node that is the canonical element of the path at the
 *  start of the operation; the operation performs a splay at i,
 *  so after the operation i is the canonical element.
 */
path PathSet::findpath(index i) { return splay(i); }

/** Return the last node in a path.
 *  @param q is the canonical element of some path
 *  @return the last node in the path containing q
 */
path PathSet::findtail(path q) {
	if (q == 0) return 0;
	while (right(q) != 0) q = right(q);
	return splay(q);
}

/** Add to the cost of every node in a path.
 *  @param q is the canonical element of some path
 *  @param x is the amount to be added to the costs of the nodes in
 *  the path
 */
void PathSet::addpathcost(path q, cost x) { dmin(q) += x; }

/** Find the the last node on a path that has minimum cost.
 *  @param q is the canonical element of some path
 *  @return a pair containing the last node on the path that has minimum
 *  cost and its cost
 */
PathSet::PathCostPair PathSet::findpathcost(path q) {
	while (1) {
		if (right(q) != 0 && dmin(right(q)) == 0)
			q = right(q);
		else if (dcost(q) > 0)
			q = left(q);
		else
			break;
	}
	q = splay(q);
	PathCostPair cp(q,dmin(q));
	return cp;
}

/** Find the root of tree represent a path, while not restructuring the tree.
 *  This is used mainly for constructing a string representation of a path.
 *  @param i is a node in some path
 *  @return the root of the search tree containing i
 */
path PathSet::findtreeroot(index i) {
	while (p(i) != 0) i = p(i);
	return i;
}

/** Join two paths at a node.
 *  @param r is the canonical element of some path
 *  @param i is an isolated node (equivalently, it is in a length 1 path)
 *  @param q is the canonical element of some path
 *  @return the new path formed by combining r,i and q (so r is the first
 *  part of the resultant path, then i, then q); this new path replaces
 *  the original paths
 */
path PathSet::join(path r, index i, path q) {
	cost dmin_i = dmin(i);
	left(i) = r; right(i) = q;
	if (r == 0 && q == 0) {
		; // do nothing
	} else if (r == 0) {
		dmin(i) = min(dmin(i),dmin(q));
		dmin(q) -= dmin(i);
		p(q) = i;
	} else if (q == 0) {
		dmin(i) = min(dmin(i),dmin(r));
		dmin(r) -= dmin(i);
		p(r) = i;
	} else {
		dmin(i) = min(dmin(r),min(dmin(i),dmin(q)));
		dmin(r) -= dmin(i); dmin(q) -= dmin(i);
		p(r) = p(q) = i;
	}
	dcost(i) = dmin_i - dmin(i);
	return i;
}

/** Divide a path at a node.
 *  @param i is a node in some path; the operation splits the path into three
 *  parts, the original portion of the path that precedes i, i itself, and
 *  the portion of the original path that follows i
 *  @return the a pair consisting of the two new path segments
 */
PathSet::PathPair PathSet::split(index i) {
	PathPair pair(0,0);

	splay(i);
	if (left(i) == 0) 
		pair.p1 = 0;
	else {
		pair.p1 = left(i); p(pair.p1) = 0; left(i) = 0;
		dmin(pair.p1) += dmin(i);
	} 
	if (right(i) == 0) 
		pair.p2 = 0;
	else {
		pair.p2 = right(i); p(pair.p2) = 0; right(i) = 0;
		dmin(pair.p2) += dmin(i);
	} 
	dmin(i) += dcost(i);
	dcost(i) = 0;

	return pair;
}

/** Determine the cost of a node without restructuring tree.
 *  This method is used mainly to construct a string representation of
 *  a path.
 *  @param i is a node in some path
 *  @return the cost of node i
 */
cost PathSet::nodeCost(index i) const {
	cost s;
	s = dcost(i);
	while (i != 0) { s += dmin(i); i = p(i); }
	return s;
}

/** Create a string representation of a path.
 *  @param q is the canonical element of some path
 *  @return the string
 */
string PathSet::path2string(path q) const {
	string s;
	if (q == 0) return s;
	s += path2string(left(q));
	s += Adt::index2string(q) + ":" + to_string(nodeCost(q)) + " ";
	s += path2string(right(q));
	return s;
}

/** Create a string representation of the tree representing a path.
 *  @param q is the canonical element of some path
 *  @return the string
 */
string PathSet::pathTree2string(path q) const {
	string s;
	if (q == 0) return s;
	bool singleton = (left(q) = 0 && right(q) == 0);
	if (!singleton) s += "(";
	s += pathTree2string(left(q));
	s += Adt::index2string(q) + ":" + to_string(nodeCost(q)) + " ";
	s += pathTree2string(right(q));
	if (!singleton) s += ")";
	return s;
}

/** Create a string representation of this object.
 *  @return the string
 */
string PathSet::toString() const {
	string s;
	for (index i = 1; i <= n(); i++) {
		if (p(i) == 0) s += path2string(i) + "\n";
	}
	return s;
}

} // ends namespace

File: js/dataStructures/misc/PathSet.mjs
/** @file PathSet.mjs
 *
 *  @author Jon Turner
 *  @date 2021
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

import Top from '../Top.mjs';
import { fassert } from '../../common/Errors.mjs';
import List from '../basic/List.mjs';
import ListSet from '../basic/ListSet.mjs';
import Scanner from '../basic/Scanner.mjs';
import SplayForest from '../trees/SplayForest.mjs';

/** Data structure representing a collection of paths.
 *
 *  Paths can be split apart or joined together and each path node
 *  has a cost. Supports efficient search for the last mincost node.
 *  in a path.
 */
export default class PathSet extends SplayForest {
	#dcost;		// #dcost[u] = cost(u)-mincost(u)
	#dmin;		// #dmin[u] = mincost(u)-mincost(p(u)) 

	/** Constructor for List object.
	 *  @param n is the range for the list
	 *  @param capacity is the max range to allocate space for
	 */
	constructor(n=10, capacity=n) {
		super(n,capacity);
		this.#dcost = new Float32Array(capacity+1).fill(0);
		this.#dmin = new Float32Array(capacity+1).fill(0);
		this.#dcost[0] = this.#dmin[0] = Infinity;
	}

	/** Assign new value to PathSet from another. 
	 *  @paran ps is a PathSet whose value is to be assigned to this
	 */
	assign(ps) {
		fassert(ps == this || !(ps instanceof PathSet));
		if (ps == this || !(ps instanceof PathSet)) return;
		super.assign(ps);
		for (let u = 1; u <= this.n; u++) {
			this.#dcost[u] = ps.#dcost[u];
			this.#dmin[u] = ps.#dmin[u];
		}
	}

	/** Assign a new value to this, by transferring contents of another list.
	 *  @param ps is a PathSet whose contents are to be transferred to this
	 */
	xfer(ps) {
		if (ps == this || !(ps instanceof PathSet)) return;
		super.xfer(ps);
		this.#dcost = ps.#dcost; ps.#dcost = null;
		this.#dmin = ps.#dmin; ps.#dmin = null;
	}
	
	/** Return to initial state */
	clear() {
		super.clear(); this.#dcost.fill(0); this.#dmin.fill(0);
	}

	/** Get/set the deltaCost of a node.
	 *  @param u is a node in a path.
	 *  @return the deltaCost of u in the tree representing the path
	 */
	dcost(u, c=-1) {
		if (c != -1) this.#dcost[u] = c;
		return this.#dcost[u];
	}
	
	/** Get/set the deltaMin of a node.
	 *  @param u is a node in a path.
	 *  @return the deltaMin of u in the tree representing the path
	 */
	dmin(u, c=-1) {
		if (c != -1) this.#dmin[u] = c;
		return this.#dmin[u];
	}
	
	/** Get/set the successor of a path.
	 *  @param p is the id of a path.
	 *  @param s is an optional new value for the successor.
	 *  @return the successor
	 */
	succ(p, s=-1) {
		if (s >= 0) this.property(p, s);
		return this.property(p);
	}
	
	/** Get the mincost of a subtree.
	 *  @param u is a node in a path.
	 *  @return the minimum cost of all nodes in subtree at u.
	 */
	mincost(u) {
		let mc = 0;
		for (let v = u; v > 0; v = this.p(v)) mc += this.dmin(v);
		return mc;
	}
	
	/** Get the cost of a node.
	 *  @param u is a node in a path.
	 *  @param mc is an optional vector with mc[u] == mincost(u);
	 *  the utility method getMincosts returns such a vector
	 *  @return the cost of u.
	 */
	cost(u,mc=null) {
		return this.dcost(u) + (mc ? mc[u] : this.mincost(u));
	}

	/** Perform a rotation in a search tree representing a path.
	 *  @param x is a node in some path; the operation performs a rotation
	 *  at the parent of x, moving x up into its parent's position.
	 */
	rotate(x) {
		let y = this.p(x); if (y <= 0) return;
		let z = this.p(y);
		let a, b, c;
		if (x == this.left(y)) {
			a = this.left(x);  b = this.right(x); c = this.right(y);
		} else {
			a = this.right(x); b = this.left(x);  c = this.left(y); 
		}
		super.rotate(x);
	
		// update dmin, dcost values
		let dmx = this.dmin(x);
		let dma = this.dmin(a); let dmb = this.dmin(b); let dmc = this.dmin(c);
			// note: if a=0, dma = Infinity; same for b,c

		this.dcost(x, this.dcost(x) + dmx);
		this.dmin(x, this.dmin(y));

		this.dmin(y, Math.min(this.dcost(y), dmb+dmx, dmc));
		this.dcost(y, this.dcost(y) - this.dmin(y))

		this.dmin(a, dma + dmx);
		this.dmin(b, dmb + dmx - this.dmin(y));
		this.dmin(c, dmc       - this.dmin(y));
	}
	
	/** Return the id of the path containing a specified node.
	 *  @param u is a node in some path
	 *  @param nosplay is an optional flag that prevents the normal
	 *  restructuring of the underlying tree
	 *  @return the id of the path after the operation is performed
	 */
	findpath(u, nosplay=false) { return super.find(u,nosplay); }
	
	/** Return the last node in a path.
	 *  @param q is the id of some path
	 *  @return the last node in the path containing q; the returned
	 *  node is also the new path id
	 */
	findtail(q) { return this.splay(super.last(q)); }
	
	/** Add to the cost of every node in a path.
	 *  @param q is the canonical element of some path
	 *  @param x is the amount to be added to the costs of the nodes in
	 *  the path
	 */
	addpathcost(q, x) { this.dmin(q, this.dmin(q) + x); }
	
	/** Find the the last node on a path that has minimum cost.
	 *  @param q is a path id
	 *  @return a pair containing the last node on the path that has minimum
	 *  cost and its cost; the returned node is also the new path id
	 */
	findpathcost(q) {
		while (true) {
			if (this.right(q) != 0 && this.dmin(this.right(q)) == 0)
				q = this.right(q);
			else if (this.dcost(q) > 0)
				q = this.left(q);
			else
				break;
		}
		q = this.splay(q);
		return [q, this.dmin(q)];
	}
	
	/** Join two paths at a node.
	 *  @param r is the canonical element of some path
	 *  @param u is an isolated node (equivalently, it is in a length 1 path)
	 *  @param q is the canonical element of some path
	 *  @return the new path formed by combining r,u and q (so r is the first
	 *  part of the resultant path, then u, then q); this new path replaces
	 *  the original paths; note: the new path inherits its successor from q
	 */
	join(r, u, q) {
		let dmin_u = this.dmin(u);
		let sq = (q ? this.succ(q) : 0);
		super.join(r,u,q);
		if (r == 0 && q == 0) {
			; // do nothing
		} else if (r == 0) {
			this.dmin(u, Math.min(this.dmin(u), this.dmin(q)));
			this.dmin(q, this.dmin(q) - this.dmin(u));
		} else if (q == 0) {
			this.dmin(u, Math.min(this.dmin(u), this.dmin(r)));
			this.dmin(r, this.dmin(r) - this.dmin(u));
		} else {
			this.dmin(u, Math.min(this.dmin(r), this.dmin(u), this.dmin(q)));
			this.dmin(r, this.dmin(r) - this.dmin(u));
			this.dmin(q, this.dmin(q) - this.dmin(u));
		}
		this.dcost(u, dmin_u - this.dmin(u)); this.succ(u, sq);
		return u;
	}
	
	/** Divide a path at a node.
	 *  @param u is a node in some path; the operation splits path into three
	 *  parts, the original portion of the path that precedes u, u itself, and
	 *  the portion of the original path that follows u
	 *  @return the a pair consisting of the two new path segments
	 *  note: the first path segment acquires u as its successor and the
	 *  second segment inherits its successor from the original
	 *  path being split
	 */
	split(u) {
		let [p,q] = super.split(u);
		let su = this.succ(u);
		if (p != 0) { this.#dmin[p] += this.dmin(u); this.p(p,0); }
		if (q != 0) { this.#dmin[q] += this.dmin(u); this.succ(q,su); }
		this.#dmin[u] += this.dcost(u); this.#dcost[u] = 0;
		return [p,q];
	}
	
	/** Compare two PathSets for equality.
	 *  @param other is the PathSet to be compared to this one
	 *  @return true if they are the same list or have the
	 *  same contents (in the same order);
	 *  they need not have the same storage capacity to be equal
	 */
	equals(other) {
		let ps = super.listEquals(other);
		if (typeof ps == 'boolean') return ps;

		let mc1 = this.getMincosts();
		let mc2 =   ps.getMincosts();
		for (let u = 1; u <= this.n; u++) {
			if (this.cost(u,mc1) != ps.cost(u,mc2)) return false;
		}
		return ps;
	}

	/** Compute the mincosts for all nodes and return an array mapping
	 *  each node to its mincost.
	 */
	getMincosts() {
		let mincost = new Float32Array(this.n+1);
		let q = new List(this.n);
		for (let r = 1; r <= this.n; r++) {
			if (this.p(r)) continue;
			q.enq(r);
			mincost[r] = this.dmin(r);
			while (!q.empty()) {
				let u = q.deq();
				if (this.left(u)) {
					let lu = this.left(u);
					q.enq(lu);
					mincost[lu] = mincost[u] + this.dmin(lu);
				}
				if (this.right(u)) {
					let ru = this.right(u);
					q.enq(ru);
					mincost[ru] = mincost[u] + this.dmin(ru);
				}
			}
		}
		return mincost;
	}

	/** Construct a string representation of this object.
	 *  @param fmt is an integer in which low bits are format options
	 *		0001 specifies that paths appear on separate lines
	 *		0010 specifies that singleton paths are shown
	 *		0100 specifies that underlying tree structure is shown
	 *		1000 specifies that dmin and dcost values are shown
	 *  @param label is an optional function used to generate the node label
	 *  @return the string
	 */
	toString(fmt=0x2, label=0) {
		let mc = this.getMincosts();
		let xlabel = (u => (label ?  this.label(u) : this.x2s(u)) +
				':' + (mc[u]+this.dcost(u)) +
				((fmt&0x8) ? `:${this.dmin(u)}:${this.dcost(u)}` : ''));
		let s = ''; let first = true;
		for (let u = 1; u <= this.n; u++) {
			if (this.p(u)) continue;
			if (this.singleton(u) && !(fmt&0x2)) continue;
			if (first) first = false;
			else s += (fmt&0x1 ? '\n' : ' ');
			s += `${this.tree2string(u,fmt,xlabel)}` +
				  (this.succ(u) ? '->' + this.x2s(this.succ(u)) : '');
		}
		return fmt&0x1 ? '{\n' + s + '\n}' : '{' + s + '}';
	}

	/** Initialize this PathSets object from a string.
	 *  @param s is a string representing a PathSets object.
	 *  @return true on success, else false
	 */
	fromString(s) {
		let sc = new Scanner(s);
		if (!sc.verify('{')) return false;
		let n = 0; let paths = []; let items = new Set();
		let cost = [];
		let prop = (u,sc) => {
						if (!sc.verify(':')) {
							cost[u] = 0; return true;
						}
						let p = sc.nextNumber();
						if (Number.isNaN(p)) return false;
						cost[u] = p;
						return true
					};
		for (let l = sc.nextIndexList('[',']',prop); l;
				 l = sc.nextIndexList('[',']',prop)) {
			for (let i of l) {
				n = Math.max(n,i);
				if (items.has(i)) return false;
				items.add(i);
			}
			let succ = 0;
			if (sc.verify('->')) {
				succ = sc.nextIndex();
				if (succ < 0) return false;
			}
			paths.push([l, succ]);
		}
		if (!sc.verify('}')) return false;
		if (n != this.n) this.reset(n);
		else this.clear();
		for (let [l,succ] of paths) {
			let p = l[0];
			this.dcost(p,0); this.dmin(p,cost[p]);
			for (let i of l) {
				if (i == p) continue;
				this.dcost(i,0); this.dmin(i,cost[i]);
				p = this.findtail(p);
				let [t1,] = this.split(p);
				p = this.join(t1,p,i);
			}
			this.succ(p,succ);
		}
		return true;
	}
}
